src/app/data-model/DecodableList.h

```diff
@@ -189,31 +162,192 @@ class DecodableMaybeFabricScopedList : public DecodableListBase, public FabricIn
+     */
+    CHIP_ERROR ComputeSize(size_t & size) const
+    {
+        if (!readerOrSpan.mIsDecodable)
+        {
+            size = readerOrSpan.mValue.mList.size();
+            return CHIP_NO_ERROR;
+        }
+        return Super::ComputeSize(size, reader());
+    }
+
+    /*
+     * @brief
+     *
+     * This call clears the TLV reader managed by this class, so it can be reused.
+     */
+    void ClearReader()
     {
-        using IteratorBase = typename detail::DecodableMaybeFabricScopedList<DataModel::IsFabricScoped<T>::value>::Iterator;
+        readerOrSpan.mValue.mHolder.mReader.Init(nullptr, 0);
+        readerOrSpan.mIsDecodable = true;
+        FabricInit();
+    }
+
+    CHIP_ERROR Decode(TLV::TLVReader & reader)
+    {
+        SetReader(reader);
+        return Super::DecodeInternal(reader);
+    }
+
+    /*
+     * @brief
+     *
+     * This call stores a TLV reader positioned on the list this class is to manage.
+     *
+     * Specifically, the passed-in reader should be pointing into the list just after
+     * having called `OpenContainer` on the list element.
+     */
+    void SetReader(const TLV::TLVReader & reader)
+    {
+        readerOrSpan.mValue.mHolder.mReader = reader;
+        readerOrSpan.mIsDecodable           = true;
+        FabricInit();
+    }
+
+    template <typename T0 = T, std::enable_if_t<DataModel::IsFabricScoped<T0>::value, bool> = true>
+    void SetFabricIndex(FabricIndex fabricIndex)
+    {
+        VerifyOrDie(!readerOrSpan.mIsDecodable);
+        readerOrSpan.mValue.mHolder.mFabricIndex.SetValue(fabricIndex);
+    }
+
+    // Iterates through all elements in the list, calling iterateFn on each element
+    // The T element passed in to iterateFn has a guaranteed lifetime of the method call,
+    // though if the implementation is a list, the element has a lifetiem beyond the call to iterateFn
+    template <class F>
+    __attribute__((always_inline)) inline CHIP_ERROR IterateGeneric(F iterateFn) const
+    {
+        if (readerOrSpan.mIsDecodable)
+        {
+            return Iterate(iterateFn);
+        }
+        else
+        {
+            return readerOrSpan.mValue.mList.Iterate(iterateFn);
+        }
+    }
+
+    // Like IterateGeneric, except only handling DecodableList, blowing up if underlying structure is List<T>
+    // Optimization which results in smaller binary size for the case we know this is a DecodableList
+    template <class F>
+    __attribute__((always_inline)) inline CHIP_ERROR Iterate(F iterateFn) const
+    {
+        Iterator iter(*this);
+        while (iter.Next())
+        {
+            bool breakLoop = false;
+            ReturnErrorOnFailure(iterateFn(iter.GetValue(), breakLoop));
+            if (breakLoop)
+            {
+                break;
+            }
+        }
+        return iter.GetStatus();
+    }
+
+    inline typename List<const T>::pointer begin() const
+    {
+        VerifyOrDie(!readerOrSpan.mIsDecodable);
```

Comment:
How does is caller know whether `begin()` is safe to call on a DecodableList they were given?

Reply:
No longer applies with the shift to not combine `Type` + `DecodableType` (code removed)